### PR TITLE
fix(topk): eliminate multi-CTA radix top-k stream hangs on SM120/SM121

### LINKED ISSUE #3610: Top-k/top-p sampler causes CUDA stream hang on consumer Blackwell GPUs (SM120/SM121)
## Bug Description

The FlashInfer top-k/top-p sampler can cause a CUDA stream to stall indefinitely on consumer Blackwell GPUs (SM120/SM121: GB10, RTX 5090, RTX 6000 Pro). Symptoms include:
- GPU utilization pinned at 96% with abnormally low power draw
- Zero token throughput
- `get_output()` spin-looping forever

Data center Blackwell GPUs (B200/B300/GB200/GB300 with SM100/SM110) appear unaffected.

## Workaround

vLLM has disabled FlashInfer by default on SM120/SM121 via [PR #44405](https://github.com/vllm-project/vllm/pull/44405). Users can still opt in with `VLLM_USE_FLASHINFER_SAMPLER=1`.

## Expected Behavior

FlashInfer should support Blackwell consumer GPUs without causing stream hangs.

## Environment

- GPU: NVIDIA GB10 (DGX Spark), RTX 5090, RTX 6000 Pro Blackwell
- vLLM version: 0.21.0+
- FlashInfer version: 0.6.6

## Related

- vLLM issue: https://github.com/vllm-project/vllm/issues/43885
- vLLM PR (workaround): https://github.com/vllm-project/vllm/pull/44405

### PR DESCRIPTION
**Pull request by waynehacking8**
> **Update — reduced to the reset-race fix only ("fix 1" / Part A).**
> Per maintainer preference, the occupancy clamp + cooperative-launch change
> ("fix 2", `RadixTopKClampGroupsToCoResidency` / `cudaLaunchCooperativeKernel`)
> was **dropped** — it changed launch semantics on every arch and was the likely
> cause of non-SM120 CI failures. The current diff is **`include/flashinfer/topk.cuh`
> only** (last-CTA-out barrier reset). The fix-2 discussion below is kept for
> historical context but is **not** in this PR.

## Description

Fixes #3610 (closes duplicates #3611, #3612): permanent sampler stream hangs on SM120/SM121 (RTX 5090 / RTX PRO 6000 / GB10) in the multi-CTA radix top-k kernels. Also the kernel behind the 12-day-old py-spy trace in vllm-project/vllm#43885 (`RadixTopKMaskLogitsMultiCTA` never completing under `CUDA_LAUNCH_BLOCKING=1`), which led vLLM to default the flashinfer sampler **off** on SM120/121 (vllm-project/vllm#44405).

Two distinct root causes in `include/flashinfer/topk.cuh`:

**1. End-of-kernel reset race (the production hang).** After the persistent row loop, the *leading* CTA cleared the group's histograms and zeroed the software barrier's `arrival_counter` — with no synchronization against peers' final `wait_ge` pass. A peer that arrived at the last barrier but whose spin-load was delayed (context time-slicing on consumer/desktop parts makes this routine) then reads `0 < target` and spins forever; the stream wedges at ~96% util / low power, matching every report in vllm#43885.
*Direct evidence:* back-to-back hammering of `top_k_mask_logits` (batch 256 × vocab 151936, fp32) wedges pre-fix **3/3 runs within ≤3400 calls**, and a side-stream dump of the live `row_states` during the wedge shows **`arrival_counter == 0` in every group while the kernel is stuck** — the reset already happened under a spinning waiter. With the launch queue drained every ≤10 calls, 60k calls never hang — exactly the "sustained load" signature from the reports.
*Fix:* `RadixGroupResetStateLastCTA` — every CTA marks its exit with a release-ordered `atom.add` on the same counter; the CTA observing the final exit is provably the last one out (a CTA only exit-marks after passing all `wait_ge` spins) and alone clears the histograms (+ deterministic scratch) and resets the counter. Replaces all three reset sites (MaskLogits / RenormProb / Unified).

**2. Co-residency never guaranteed (latent, deterministic under SM pressure).** Grid sizing assumed 1 CTA/SM on an idle device; a plain launch guarantees nothing. With a burner kernel pinning SMs, the deadlock threshold lands exactly at free SMs < `ctas_per_group` (= 7 here: SM120's 99KB smem/block makes groups 2-4× wider than on SM100's 227KB, which is why these archs live closest to the edge). Two such kernels on different streams can starve each other permanently, and the `num_groups==0 → 1` fallback could launch `ctas_per_group > num_sms` grids that deadlock on an *idle* GPU.
*Fix:* `RadixTopKClampGroupsToCoResidency` (`cudaOccupancyMaxActiveBlocksPerMultiprocessor`-verified clamp; hard `cudaErrorCooperativeLaunchTooLarge` instead of a guaranteed hang when one group can't co-reside) + `LaunchRadixTopKMultiCTAKernel` (gang-scheduled `cudaLaunchCooperativeKernel`, falling back to plain launch when the device lacks coop launch **or the stream is capturing** — CUDA-graph paths keep today's behavior and still get fix 1).

## Validation (RTX PRO 6000, SM120, 188 SMs, CUDA 13.0)

| Scenario | pre-fix | post-fix |
|---|---|---|
| back-to-back hammer, 20k calls b256×v151936 | **wedged 3/3 (≤3400 calls)** | **0 hangs in 120k calls (6/6 runs)** |
| 2 streams, mask_logits + renorm_probs ×2000 | **wedged** | completes 1.6s, 6/6 |
| burner holding 182/188 SMs (free < group) | spins until burner exits | gang launch waits, completes |
| correctness (1..256 × 151936/152064/262144, vs torch.topk) | match | match |
| `tests/utils/test_topk.py` + `test_sampling.py -k "top_k or topk"` | — | **1661 passed** |
| perf (gpu ms/call, 4 shapes) | 0.024/0.050/0.365/0.118 | 0.042/0.055/0.356/0.067 (noise-level; +~1µs host launch) |

**Ablation for reviewers:** the reset fix *alone* clears every observed wedge (hammer 3/3, 2-stream 3/3 with coop launch toggled off). The occupancy clamp + cooperative launch close the synthetic-but-real starvation class and the idle-GPU `ctas_per_group > num_sms` case. Trade-off to weigh: gang scheduling waits for full residency instead of cascading (the 181-burner case waits ~3.8s rather than trickling at 4.8ms) — if that's a concern for heavily shared GPUs, fix 1 + the clamp's hard-error can ship standalone; happy to split.

Note for vLLM: once this lands, the SM120/121 default-off in vllm#44405 can be revisited.

## Tests

- [x] Tests pass locally (1661 passed; full repro matrix above; repro scripts available on request)
- [ ] New unit test — the hang requires sustained concurrent launch pressure that doesn't fit the unit-test harness; validation is via the deterministic repro scripts documented above. Open to adding a stress test under an opt-in marker if maintainers want it in-tree.

## Disclosure

AI-assisted (Claude): investigation, fix, and validation runs; the submitter reviewed every line and the memory-ordering argument.

## Summary by CodeRabbit

* **Improvements**
  * Enhanced multi-CTA Top-K kernels with more reliable end-of-round coordination, improving stability and throughput during synchronized histogram/state resets.
  * Refined host-side multi-CTA launch behavior to improve cooperative launch selection and robustness.

* **Bug Fixes**
  * Prevented potential race conditions and deadlocks by ensuring the final exiting group performs the required cross-CTA reset work, avoiding issues when compute groups are only partially resident.

### CODE REVIEW

File: include/flashinfer/topk.cuh [hunk 1]
```diff
@@ -161,6 +243,57 @@ __device__ __forceinline__ void AdvanceRadixGroupBarrier(RadixRowState* state, i
...
+ * \param ctas_per_group Number of CTAs participating in the group
+ * \param tx Thread index within the block
+ */
+template <uint32_t BLOCK_THREADS>
+__device__ __forceinline__ void RadixGroupResetStateLastCTA(
+    RadixRowState* state, RadixDeterministicCollectScratch* det_scratch, int barrier_phase,
+    uint32_t ctas_per_group, uint32_t tx) {
+  constexpr uint32_t RADIX = 256;
+  __shared__ int s_is_last_cta;
+  if (tx == 0) {
```
[the remark this answers is not in the document]
Comment (waynehacking8) [reply include/flashinfer/topk.cuh:213]: Fair scrutiny — addressed in 067d37c with an explicit `__syncthreads()` immediately before the exit mark, so tx0's `fence.acq_rel.gpu` cumulativity provably covers every thread's prior writes. Two notes for the record: (1) this syncthreads + thread-0 release-fence idiom is the file's established convention — `red_release`/`st_release` document exactly this ('the fence also releases data that was weakly-written by other threads prior to the last syncthreads') and `AdvanceRadixGroupBarrier` already relies on it; (2) strictly, the exit mark itself is an execution-ordering protocol (the last CTA needs to know peers have *passed their final spin*, not observe their data), and the state clears that do need publication were already ordered by the existing syncthreads before the counter reset — but the extra convergence point makes the argument airtight at zero cost (callers are converged at that point anyway). Re-validated post-change: 20k back-to-back hammer clean, correctness vs torch.topk holds.

File: include/flashinfer/topk.cuh [hunk 2]
```diff
@@ -101,6 +101,25 @@ __device__ __forceinline__ void st_release(int* ptr, int val) {
...
+  int old_val = 0;
+#if (__CUDA_ARCH__ >= 700)
+  // SM70 and newer use memory consistency qualifiers
+  // Release pattern using acq_rel fence + relaxed modifier
+  // (The fence also releases data that was weakly-written by other threads prior to the last
+  // syncthreads)
+  asm volatile("fence.acq_rel.gpu;\n");
+  asm volatile("atom.relaxed.gpu.global.add.s32 %0, [%1], %2;\n"
+               : "=r"(old_val)
+               : "l"(ptr), "r"(val));
```
[the remark this answers is not in the document]
Comment (waynehacking8) [reply include/flashinfer/topk.cuh]: Done in 067d37c — `"memory"` clobbers on both asm statements of `atom_add_release`. (The pre-existing `red_release`/`st_release`/`ld_acquire` share the unclobbered pattern; left untouched here to keep the diff minimal, but happy to harden all of them in this PR or a follow-up if you prefer.)

File: include/flashinfer/topk.cuh [hunk 3]
```diff
@@ -138,6 +157,69 @@ inline RadixDeterministicCollectScratch* MaybeGetRadixDeterministicCollectScratc
...
+                                                 void** args, size_t smem_size,
+                                                 cudaStream_t stream) {
+  int device = 0;
+  FLASHINFER_CUDA_CALL(cudaGetDevice(&device));
+  int supports_coop_launch = 0;
+  FLASHINFER_CUDA_CALL(
+      cudaDeviceGetAttribute(&supports_coop_launch, cudaDevAttrCooperativeLaunch, device));
+  cudaStreamCaptureStatus capture_status = cudaStreamCaptureStatusNone;
+  FLASHINFER_CUDA_CALL(cudaStreamIsCapturing(stream, &capture_status));
+  if (supports_coop_launch && capture_status == cudaStreamCaptureStatusNone) {
```
[the remark this answers is not in the document]
Comment (waynehacking8) [reply include/flashinfer/topk.cuh]: Done in 067d37c — `cudaDevAttrCooperativeLaunch` is now cached per device (zero-initialized atomic array, relaxed ordering; benign double-query race at worst). `cudaStreamIsCapturing` stays per-call since capture state is dynamic.

**Comment by waynehacking8** [comment]
@flashinfer-bot's pipeline link is internal — could a maintainer share which jobs/configs failed (names + key log lines)? Local state for reference: 1661 upstream tests pass on SM120, 120k-launch stress clean, and the PR description includes an ablation showing the **last-CTA-out reset fix alone clears all observed hangs** — so if the failures cluster on the cooperative-launch or occupancy-clamp half (other archs / launch-config edge cases I can't exercise here), I can split that out and ship the reset fix + hard-error clamp standalone immediately. Happy to iterate fast on whatever the logs show.

**Comment by waynehacking8** [comment]
@kahyunnam Thanks for triggering the pipeline. While we wait on the failing-job details, one concrete candidate surfaced during #3625's validation that could explain CI failures in concurrent configurations: **the per-launch `cudaFuncSetAttribute(MaxDynamicSharedMemorySize)` on the shared kernel function races across host threads** (pre-existing on main, not introduced here) — when two threads/streams drive the multi-CTA launchers with different vocab sizes, the attribute flips mid-launch and surfaces as `invalid argument` or `too many blocks in cooperative launch`. The cooperative launch added in this PR makes that pre-existing race *visible* (a plain launch with a stale-but-larger attribute silently works; a coop launch validates and fails). If the failing jobs show those two error strings, that's the mechanism — and the fix is small (set the attribute once per (kernel, max-smem) instead of per launch); I can fold it into this PR. If the failures are something else entirely, the log lines will tell. Also note the ablation option stands: reset-fix + clamp without coop launch passes everything on my side.

**Comment by waynehacking8** [comment]
@kahyunnam Thanks for running `tests/utils`. The pipeline shows 6/20 passed, but I can't see the internal GitLab logs (the `nv/` pipeline is NVIDIA-internal), so I can't tell which tests are failing — could you share the failing test names or a log excerpt?

For reference, on an RTX PRO 6000 (SM120) the topk path passes locally: `tests/utils/test_topk.py` + `test_sampling.py` (`-k "top_k or topk"`) = 1661 passed / 849 skipped, plus a 120k back-to-back hammer with zero hangs (the original repro wedged ≤3400 calls pre-fix). The latest commit (`067d37c8`) addressed the review flags: `__syncthreads()` before the exit-mark release atomic so peer threads' output writes are in the release sequence, `"memory"` clobbers on the fence/atomic pair, and the cooperative-launch attribute is now cached per-device. Happy to fix whatever's failing on the H100 / A10G / T4 runners — just need to see what broke.

**Comment by waynehacking8** [comment]
I've pushed a reduction (`9cf8fa7`) that should clear the `tests/utils` failures.

Since I can't see the internal pipeline logs and everything passes on SM120, I split the PR by blast radius and kept only the half that is the actually-reported hang and is architecture-neutral:

**Kept — the end-of-kernel barrier reset race (root cause #1, the observed production hang).** The arrival counter is reset by the *last* CTA to exit the group (release-ordered atomic exit mark) instead of the leading CTA, so a peer still spinning in its final `wait_ge` can't be stranded. This is the defect behind #3610/#3611/#3612 and vllm-project/vllm#43885, and the ablation in the PR shows it alone clears every observed hang.

**Dropped — the co-residency hardening (root cause #2 mitigation):** `RadixTopKClampGroupsToCoResidency` (which returns `cudaErrorCooperativeLaunchTooLarge` when a group can't be fully co-resident) and the `LaunchRadixTopKMultiCTAKernel` cooperative-launch wrapper. These change launch semantics on *every* arch — they turn previously-working oversubscribed grids into hard errors and switch the launch path to `cudaLaunchCooperativeKernel` — which is the most likely reason `tests/utils` regressed on T4/A10G/H100 but not on SM120. The five multi-CTA launchers now revert to the plain `cudaLaunchKernel` they used on `main`, so that surface is no worse than today's `main`.

Net diff is now **91/-47** (was 218/-50); the filed hang is fixed and everything else returns to `main`'s proven behavior. I'm happy to bring back the stronger gang-scheduling guarantee as a separate, arch-validated PR.

**Validation** (RTX PRO 6000, SM120, CUDA 13.0, `FLASHINFER_CUDA_ARCH_LIST=12.0f`):
- JIT build for `120f`: OK
- `tests/utils/test_topk.py`: **617 passed, 0 failed** (729 arch/size-skipped)
- 3000-launch multi-CTA hammer (vocab=256k) under `CUDA_LAUNCH_BLOCKING=1`: no hang; per-row finite counts exactly equal `k`

@kahyunnam — when you have a moment, could you re-run `/bot run tests/utils` on the reduced version? If specific configs still fail I can iterate, but this removes the launch-path change that I believe was the culprit.

**Comment by waynehacking8** [comment]
Good catch — fixed in `2c8c824`. `<array>` and `<atomic>` were added only for the cooperative-launch wrapper's `std::array<std::atomic<int>>` per-device cache, which the reduction removed. Confirmed no remaining `std::` usage in the file (the `atom_add_release`/exch helpers are CUDA built-ins / inline PTX), reverted to `main`'s include set, and rebuilt the topk module clean for SM120 (`120f`).

**Comment by waynehacking8** [comment]
@kahyunnam thanks for kicking off the pipeline. Since the GitLab pipeline is internal (I can't see which jobs failed), here's what I can verify on my side.

I ran `tests/utils/test_topk.py` on an **SM120 (RTX PRO 6000)** with this branch's `topk.cuh`:

```
617 passed, 729 skipped, 0 failed, no hang (105s)
```

So the multi-CTA radix top-k path this PR touches passes on the arch the fix targets, and it does **not** reproduce the SM120/121 stream hang (#3610) that motivated the PR. The public CI here is also green on the substantive jobs (H100 + A10G x5 JIT Unittest, AOT build/import across cu126–cu130); the only red marks are `Deploy Docs` (a pre-existing `flashinfer.comm` autosummary import error, unrelated to topk) and the governance `remove-label`/T4-fast-fail (the T4 `JIT Rerun` passes).

Could you share which jobs failed in pipeline #54863357? I can't open the internal link, so I can't tell whether any of the 10 are actually related to this change vs. infra/other suites — happy to dig into any that are.

**Comment by kahyunnam** [comment]
@waynehacking8 thanks for the due diligence here -- I checked the CI pipeline logs for the failing jobs, these seem like CI issues (hardware job timeout). The test results look good to me!

**Review by kahyunnam — APPROVED** [review]
LGTM, thanks for the contribution
